aesd-char-driver/main.c

```diff
@@ -96,6 +98,120 @@ ssize_t aesd_read(struct file *filp, char __user *buf, size_t count,
+            ret_value = mutex_lock_interruptible(&dev->buffer_lock);
+            if(ret_value !=0)
+            {
+                ret_value = -ERESTART;
+                PDEBUG("Error: Unable to do mutex lock");
+                goto exit;
+            }
+            //Calculate the total length 
+            for(int i= dev->buffer.out_offs; i!=dev->buffer.in_offs;)
+            {
+                total_length +=dev->buffer.entry[i].size;
+                i = (i+1)%AESDCHAR_MAX_WRITE_OPERATIONS_SUPPORTED;
+            }
+            mutex_unlock(&dev->buffer_lock);
+            ret_value = total_length-1+offset;
+            break;
+        default:
+            ret_value = -EINVAL;
+            goto exit;           
+    }
+    if(ret_value<0 )
+    {
+        ret_value = -EINVAL;
+        goto exit;
+    }
+
+   
+    filp->f_pos = ret_value;
+    PDEBUG("File position seeked to %d",filp->f_pos);
+
+exit:
+    return ret_value;
+}
+
+
+long aesd_unlocked_ioctl(struct file *filp, unsigned int cmd, unsigned long arg) 
+{
+     struct aesd_dev *dev = filp->private_data;
+    long ret_value = 0;
+    struct aesd_seekto seek_params;
+    int total_length = 0;
+    PDEBUG("Inside aesd_unlocked_ioctl");
+    //Check if the command is AESDCHAR_IOCSEEKTO
+    if (cmd != AESDCHAR_IOCSEEKTO)
+    {
+        PDEBUG("Error: aesd_unlocked_ioctl Invalid inputs\n");
+        ret_value = -ENOTTY;
+        goto exit;
+
+    }
+    //Copy from user space to kernel space
+    if (copy_from_user(&seek_params, (struct aesd_seekto __user *)arg, sizeof(seek_params)))
+    {
+        PDEBUG("Error: Copying from user failed\n");
+        ret_value = -EFAULT;
+        goto exit;
+    }
+    //Check if write_cmd exceeds max buffer operations
+    if(seek_params.write_cmd > AESDCHAR_MAX_WRITE_OPERATIONS_SUPPORTED)
+    {
+        PDEBUG("Error: Invalid command index offset\n");
+        ret_value = -EINVAL;
+        goto exit; 
+    }
+
+    seek_params.write_cmd = (seek_params.write_cmd+dev->buffer.out_offs)% AESDCHAR_MAX_WRITE_OPERATIONS_SUPPORTED;
+    PDEBUG("Write cmd is %d, write cmd offset is %d\n",seek_params.write_cmd,seek_params.write_cmd_offset);
+    if(seek_params.write_cmd_offset > dev->buffer.entry[seek_params.write_cmd].size)
+    {
+        PDEBUG("Error: Invalid command command offset\n");
+        ret_value = -EINVAL;
+        goto exit; 
+    }
+    //Lock mutex
+    ret_value = mutex_lock_interruptible(&dev->buffer_lock);
+    if(ret_value !=0)
+    {
+        PDEBUG("Error: Unable to acquire mutex lock\n");
+        ret_value = -ERESTART;
+        goto exit;
+    }
+    // Calculate the total length
+    for(int i=dev->buffer.out_offs;i!=seek_params.write_cmd;)
+    {
+        total_length+=dev->buffer.entry[i].size;
+        i = (i+1)%AESDCHAR_MAX_WRITE_OPERATIONS_SUPPORTED;
+    }
+    filp->f_pos = total_length + seek_params.write_cmd_offset;
+    mutex_unlock(&dev->buffer_lock);
```

Comment:
Could move this up one line, no need to lock f_pos or seek_params.

Reply:
Hi Dan, I somehow missed your first comment earlier and now I am trying to update with assignment9 runner. I had missed to push the merged changes with assignment 9. However when I try to rerun the job, i could see the "https://buildroot.busybox.net/" server is down because of which my actions runner is not getting initiated.  Do you have any suggestions on this regard on how to overcome?
![image](https://github.com/user-attachments/assets/8935f6e0-791e-43bf-9344-ec0072537c20)